src/coreclr/jit/valuenum.cpp

```diff
@@ -15034,3 +15081,149 @@ void ValueNumStore::PeelOffsets(ValueNum* vn, target_ssize_t* offset)
+
+//--------------------------------------------------------------------------------
+// VNWalkPhisInternal: internal worker for VNWalkPhis, see comments in VNWalkPhis
+//
+template <typename TPhiArgVisitorFunc>
+ValueNumStore::PhiDefWalkResult ValueNumStore::VNWalkPhisInternal(ValueNum           phiDefVN,
+                                                                  VNFuncApp          phiDefFunc,
+                                                                  TPhiArgVisitorFunc phiArgVisitor,
+                                                                  int                maxDepth,
+                                                                  ValueNumKind       vnKind,
+                                                                  int&               realVNsVisited,
+                                                                  SmallValueNumSet&  hashSet)
+{
+    assert((maxDepth == -1) || (maxDepth > 0));
+    assert((phiDefFunc.m_func == VNF_PhiDef) || (phiDefFunc.m_func == VNF_PhiMemoryDef));
+
+    const bool isMemory = (phiDefFunc.m_func == VNF_PhiMemoryDef);
+
+    // VNF_PhiDef(LclNum, SsaNum, VNF_Phi)
+    // VNF_PhiMemoryDef(BB, VNF_Phi)
+    ValueNum nextPhiArg = isMemory ? phiDefFunc.m_args[1] : phiDefFunc.m_args[2];
+    if (!isMemory && (phiDefFunc.m_args[0] == BAD_VAR_NUM))
+    {
+        JITDUMP("We can't process this PhiDef - bail out.\n");
+        return PhiDefWalkResult::InvalidPhiDef;
+    }
+
+    assert(!hashSet.Lookup(phiDefVN));
+    hashSet.Add(m_pComp, phiDefVN);
+
+    // Now we iterate over Phi arguments, e.g.
+    //  2 args: VNF_Phi(SSA1, SSA2)
+    //  3 args: VNF_Phi(SSA1, VNF_Phi(SSA2, SSA3))
+    //  4 args: VNF_Phi(SSA1, VNF_Phi(SSA2, VNF_Phi(SSA3, SSA4)))
+    // etc.
+    //
+    while (nextPhiArg != NoVN)
+    {
+        ValueNum  current = nextPhiArg;
+        VNFuncApp phiArgFuncApp;
+
+        // nextPhiArg is either VNF_Phi or SSA num (last)
+        if (GetVNFunc(nextPhiArg, &phiArgFuncApp))
+        {
+            assert(phiArgFuncApp.m_func == VNF_Phi);
+            current    = phiArgFuncApp.m_args[0];
+            nextPhiArg = phiArgFuncApp.m_args[1];
+        }
+        else
+        {
+            // Ok this was the last Phi arg
+            nextPhiArg = NoVN;
+        }
+
+        // We expect a constant VN here representing the SSA number
+        assert(IsVNConstant(current));
+
+        // Extract the real VN from the Phi arg
+        unsigned phiArgSsaNum = ConstantValue<unsigned>(current);
+
+        ValueNum phiArgVN;
+        if (isMemory)
+        {
+            phiArgVN = m_pComp->GetMemoryPerSsaData(phiArgSsaNum)->m_vnPair.Get(vnKind);
+        }
+        else
+        {
+            // NOTE: it's unfortunate that LclNum is not a VN.
+            unsigned phiDefLclNum = phiDefFunc.m_args[0];
+            phiArgVN              = m_pComp->lvaTable[phiDefLclNum].GetPerSsaData(phiArgSsaNum)->m_vnPair.Get(vnKind);
+        }
+
+        if (hashSet.Lookup(phiArgVN))
```

Comment:
It looks like we will always will end up adding this to the hash map in the PHI case. Can we instead make `SmallValueNumSet::Add` return false if it was already in the set to avoid avoid the duplicate lookup?

Reply:
Also, consider using `ArrayStack` instead of recursion